src/glue/jobs/json_s3_to_parquet.py

```diff
@@ -132,105 +141,246 @@ def add_index_to_table(table_key, table_name, processed_tables, unprocessed_tabl
                 if succinct_name in df_with_index.columns:
                     continue
-                df_with_index = df_with_index.withColumnRenamed(
-                        c, succinct_name)
+                df_with_index = df_with_index.withColumnRenamed(c, succinct_name)
     return df_with_index
 
+
+def cast_glue_data_types(
+    dynamic_frame: DynamicFrame,
+    glue_database_name: str,
+    glue_table_name: str,
+) -> DynamicFrame:
+    """Takes a DynamicFrame and Glue table name and force Glue to respect our data
+    type choices when resolving types from a DynamicFrame object to a Spark Dataframe.
+    This
+
+    Args:
+        dynamic_frame (awsglue.DynamicFrame): DynamicFrame object input
+        glue_database_name (str): Name of the database to pull the glue table from
+        glue_table_name (str): Name of the glue table to pull the json schema from
+
+    Returns:
+        awsglue.DynamicFrame: returns a DynamicFrame with those resolved data types
+    """
+    # Load the input glue data
+    glue_client = boto3.client("glue")
+    response = glue_client.get_table(
+        DatabaseName=glue_database_name, Name=glue_table_name
+    )
+    table = response["Table"]
+    glue_table_schema = table["StorageDescriptor"]["Columns"]
+    specs = []
+    for field_info in glue_table_schema:
+        json_schema = parse_hive_schema(
+            hive_str=field_info["Type"], top_level_field=field_info["Name"]
+        )
+        specs.extend(convert_json_schema_to_specs(json_schema))
+    glue_table = dynamic_frame.resolveChoice(specs=specs)
+    return glue_table
+
+
+def parse_hive_schema(hive_str: str, top_level_field: str) -> dict:
+    """
+    Expands embedded Hive struct strings to Python dictionaries
+
+    Example)
+    Hive struct:
+        arrayofobjectsfield: array<struct<filename:string,timestamp:string>>
+    dict struct:
+        {'arrayofobjectsfield': [{'filename': 'string', 'timestamp': 'string'}]}
+
+    Args:
+        hive_str (str): string form of the hive schema struct
+        top_level_field (str): the field that contains this hive string
+
+    Returns:
+        dict: python dictionary format of input hive struct
+    """
+    r = re.compile(r"(.*?)(struct<|array<|[:,>])(.*)")
+    root = dict()
+
+    # adjust hive string to have top level field in the hive str
+    hive_str = f"struct<{top_level_field}:{hive_str}>"
+
+    to_parse = hive_str
+    parents = []
+    curr_elem = root
+
+    key = None
+
+    iteration = 0
+    while to_parse:
+        if iteration > MAX_ITERATIONS:
+            raise Exception(
+                f"Parsing hive schema reached maximum iterations of {MAX_ITERATIONS}"
+            )
+        left, operator, to_parse = r.match(to_parse).groups()
+
+        if operator == "struct<" or operator == "array<":
+            parents.append(curr_elem)
+            new_elem = dict() if operator == "struct<" else list()
```

Comment:
I'm a little confused here, you set the key and then you write over the existing dict with an empty one?